### DISCUSSION TITLE
psqldef does not preserve DEFAULT from function calls

### DISCUSSION THREAD

**Discussion by alecthomas** [sqldef/sqldef#622] [Feature requests]
Specifically with the uuid-ossp extension this fails:

```
  "key" UUID NOT NULL DEFAULT uuid_generate_v4(),
```

## Platform

* OS: macOS
* RDBMS: PostgreSQL
* Version: 0.15.22

## --export output

```sql
CREATE EXTENSION "uuid-ossp";

CREATE TABLE "public"."artefacts" (
    "id" serial NOT NULL,
    "digest" character varying(64) NOT NULL,
    "path" character varying(128) NOT NULL,
    "uri" character varying(4096) NOT NULL,
    PRIMARY KEY ("id")
);

CREATE TABLE "public"."deployment_artefacts" (
    "deployment_id" integer NOT NULL,
    "artefact_id" integer NOT NULL,
    PRIMARY KEY ("deployment_id", "artefact_id")
);

ALTER TABLE ONLY "public"."deployment_artefacts" ADD CONSTRAINT "deployment_artefacts_artefact_id_fkey" FOREIGN KEY ("artefact_id") REFERENCES "public"."artefacts" ("id") ON UPDATE NO ACTION ON DELETE CASCADE;

ALTER TABLE ONLY "public"."deployment_artefacts" ADD CONSTRAINT "deployment_artefacts_deployment_id_fkey" FOREIGN KEY ("deployment_id") REFERENCES "public"."deployments" ("id") ON UPDATE NO ACTION ON DELETE CASCADE;

CREATE TABLE "public"."deployments" (
    "id" serial NOT NULL,
    "module_id" integer NOT NULL,
    "schema" text NOT NULL,
    PRIMARY KEY ("id")
);

ALTER TABLE ONLY "public"."deployments" ADD CONSTRAINT "deployments_module_id_fkey" FOREIGN KEY ("module_id") REFERENCES "public"."modules" ("id") ON UPDATE NO ACTION ON DELETE CASCADE;

CREATE TABLE "public"."modules" (
    "id" serial NOT NULL,
    "name" character varying(128) NOT NULL,
    PRIMARY KEY ("id")
);
```

## Input SQL

```sql
CREATE EXTENSION IF NOT EXISTS "uuid-ossp";

CREATE TABLE modules (
  id SERIAL PRIMARY KEY, 
  name VARCHAR(128) NOT NULL
);

CREATE TABLE deployments (
  id SERIAL PRIMARY KEY,
  "key" UUID NOT NULL DEFAULT uuid_generate_v4(),
  module_id INTEGER NOT NULL REFERENCES modules(id) ON DELETE CASCADE,
  "schema" TEXT NOT NULL
);

CREATE TABLE artefacts (
  id SERIAL PRIMARY KEY,
  -- SHA256 digest of content.
  digest VARCHAR(64) NOT NULL,
  -- Path relative to the module root.
  path VARCHAR(128) NOT NULL,
  -- Absolute URI to the stored artefact.
  uri VARCHAR(4096) NOT NULL
);

CREATE TABLE deployment_artefacts (
  deployment_id INTEGER NOT NULL REFERENCES deployments(id) ON DELETE CASCADE,
  artefact_id INTEGER NOT NULL REFERENCES artefacts(id) ON DELETE CASCADE,
  PRIMARY KEY (deployment_id, artefact_id)
);
```

## Current output

```
🐚 ~/Projects/ftl $ psqldef -W secret --dry-run ftl < backplane/sql/schema/000_init.sql                                                                       main
-- dry run --
ALTER TABLE "public"."deployments" ADD COLUMN "key" uuid NOT NULL DEFAULT 'uuid_generate_v4';
```

## Expected output

```
🐚 ~/Projects/ftl $ psqldef -W secret --dry-run ftl < backplane/sql/schema/000_init.sql                                                                       main
-- dry run --
ALTER TABLE "public"."deployments" ADD COLUMN "key" uuid NOT NULL DEFAULT uuid_generate_v4();
```

PS. Love this tool and in particular the simplicity. It's awesome.

**Comment by k0kubun** [comment]
Thanks for reporting this. It seems that the generic parser of sqldef currently parses `uuid_generate_v4()` as an `SQLVal` of `StrVal` `'uuid_generate_v4'`.

For psqldef, we're in the middle of migrating the generic parser, which supports all database types, to the PostgreSQL-specific parser, which is backed by the actual PostgreSQL parser. Instead of fixing the generic parser, we want to let the PostgreSQL-specific parser support the table definition you want to use and let it generate a function-call node as a default value.

Would you be interested in working on a PR to fix it? The new parser code lives in `database/postgres/parser.go` (not `parser/parser.y`) and we want to parse `DEFAULT uuid_generate_v4()` (and anything else you currently use) with it.

**Comment by S-YOU** [comment]
I took a look  postgres/parser.go, and there is code for parsing function and its argument properly, but probably not being used.
Any idea why? @k0kubun 

https://github.com/k0kubun/sqldef/blob/master/database/postgres/parser.go#L332-L349

**Comment by k0kubun** [comment]
Sorry, the link seems outdated today. Can you post a file link with a precise commit hash instead of `master`? You can press `y` on that page to update the link.

**Comment by S-YOU** [comment]
> Sorry, the link seems outdated today. Can you post a file link with a precise commit hash instead of master? You can press y on that page to update the link.

Sorry, forgot to use commit hash. It was FuncCall part here.

https://github.com/k0kubun/sqldef/blob/b7330fab3f28e431d22c225de89276cc7423cf6f/database/postgres/parser.go#L358-L375

**Comment by k0kubun** [comment]
That line was introduced for https://github.com/k0kubun/sqldef/commit/fb6ef419. I'm not sure what you mean by "probably not being used". It doesn't seem like a dead path.

Can you try dumping the source tree by `pp.Println` using https://github.com/k0kubun/pp (it's also nice to use `SetExportOnly`)? I bet `pgquery.Node_FuncCall` is there, but its parent or ancestor may not be handled by the parser.

**Comment by S-YOU** [comment]
> That line was introduced for https://github.com/k0kubun/sqldef/commit/fb6ef419. I'm not sure what you mean by "probably not being used". It doesn't seem like a dead path.

If that path is being used, alter table statement should not be treated `uuid_generate_v4` as string, isn't it? that part seems to be parsing function call properly. And it is working if the field is included during create table phase. Only alter table is not working.

```
-- dry run --
ALTER TABLE "public"."deployments" ADD COLUMN "key" uuid NOT NULL DEFAULT 'uuid_generate_v4';
```

OP has the repro and here is minium repro.
1. create this table first
```
CREATE TABLE deployments (
	id SERIAL PRIMARY KEY,
	"schema" TEXT NOT NULL
);
```

2. then add key 
```
CREATE TABLE deployments (
	id SERIAL PRIMARY KEY,
	"key" UUID NOT NULL DEFAULT uuid_generate_v4(),
	"schema" TEXT NOT NULL
);
```

> Can you try dumping the source tree by pp.Println using https://github.com/k0kubun/pp (it's also nice to use SetExportOnly)? I bet pgquery.Node_FuncCall is there, but its parent or ancestor may not be handled by the parser.

I don't to how to run sqldef from source code, I will take a look.

**Comment by k0kubun** [comment]
> If that path is being used, alter table statement should not be treated uuid_generate_v4 as string, isn't it? that part seems to be parsing function call properly. And it is working if the field is included during create table phase. Only alter table is not working.

Yeah. It'd be nice to double-check what AST it's currently generating, but it might be falling under a different rule.

> OP has the repro and here is minium repro.

Nice :+1:

> I don't to how to run sqldef from source code, I will take a look.

Thank you :pray: Normally, I first start databases with `docker-compose up`, and then run psqldef with `go run cmd/psqldef/psqldef.go -Upostgres -h 127.0.0.1 test < schema.sql`.
